### PR TITLE
Expose turbo module registry API

### PR DESCRIPTION
**Pull request by ZihanChen-MSFT**
- Add `AddTurboModule` to `IReactPackageBuilder.idl`.
- Pass it all the way to `facebook::react::CreateReactInstance` in `ReactInstanceWin`.
- `Clipboard` registers itself as a turbo module instead of a native module, only when web debugger is disabled.
- `TurboModulesProvider` has limited functionalities, which only satisfy `Clipboard` module's need.
- `Clipboard` module is updated to run Clipboard functions in UI thread.
- `Clipboard` is updated to run `ReactPromise` in Javascript Engine's thread.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5085)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/IReactPackageBuilder.idl [hunk 1]
```diff
@@ -14,5 +14,6 @@ namespace Microsoft.ReactNative {
   interface IReactPackageBuilder {
     void AddModule(String moduleName, ReactModuleProvider moduleProvider);
     void AddViewManager(String viewManagerName, ReactViewManagerProvider viewManagerProvider);
+    void AddTurboModule(String moduleName, ReactModuleProvider moduleProvider);
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/IReactPackageBuilder.idl:17]: this is a breaking change. Apps depend on this interface.
Moreover this will break creating new apps as the template hasn't been updated to generate this new method.

Comment (vmoroz) [reply vnext/Microsoft.ReactNative/IReactPackageBuilder.idl:17]: This interface is implemented by us. No apps are going to break. No new code is going to affected too. This is the power of the 'Builder' pattern. Applications must implement IReactPackageProvider with a single method that receives the IReactPackageBuilder as a parameter.

---
In reply to: [433524380](https://github.com/microsoft/react-native-windows/pull/5085#discussion_r433524380) [](ancestors = 433524380)

File: vnext/Microsoft.ReactNative/ReactPackageBuilder.h [hunk 2]
```diff
@@ -11,15 +12,18 @@ namespace winrt::Microsoft::ReactNative {
 struct ReactPackageBuilder : winrt::implements<ReactPackageBuilder, IReactPackageBuilder> {
   ReactPackageBuilder(
       std::shared_ptr<NativeModulesProvider> const &modulesProvider,
-      std::shared_ptr<ViewManagersProvider> const &viewManagersProvider) noexcept;
+      std::shared_ptr<ViewManagersProvider> const &viewManagersProvider,
+      std::shared_ptr<TurboModulesProvider> const &turboModulesProvider) noexcept;
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/ReactPackageBuilder.h:16]: I think this is a publicly-visible breaking change too?

File: vnext/Microsoft.ReactNative/IReactPackageBuilder.idl (hunk 1 above)
Comment (ZihanChen-MSFT) [reply vnext/Microsoft.ReactNative/IReactPackageBuilder.idl:17]: What does `the template hasn't been updated` mean? In my understanding I have already provided implementation to this new function. Is there any other place out side of Microsoft.ReactNative that also implements this interface?

File: vnext/Microsoft.ReactNative/TurboModulesProvider.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/TurboModulesProvider.h:1]: >All rights reserved. [](start = 40, length = 20)

Please do not add 'All rights reserved.' It is deleted by PR #5083.

File: vnext/Microsoft.ReactNative/TurboModulesProvider.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  virtual std::shared_ptr<facebook::react::TurboModule> getModule(
+      const std::string &moduleName,
+      const std::shared_ptr<facebook::react::CallInvoker> &callInvoker) noexcept override;
+  virtual std::vector<std::string> getEagerInitModuleNames() noexcept override;
+
+ public:
+  void AddModuleProvider(winrt::hstring const &moduleName, ReactModuleProvider const &moduleProvider) noexcept;
+
+ private:
+  std::map<std::string, ReactModuleProvider> m_moduleProviders;
```
Comment (asklar) [inline vnext/Microsoft.ReactNative/TurboModulesProvider.h:20]: avoid use of std::map

File: vnext/Microsoft.ReactNative/TurboModulesProvider.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+-------------------------------------------------------------------------------*/
+std::shared_ptr<facebook::react::TurboModule> TurboModulesProvider::getModule(
+    const std::string &moduleName,
+    const std::shared_ptr<facebook::react::CallInvoker> &callInvoker) noexcept {
+  auto it = m_moduleProviders.find(moduleName);
+  if (it == m_moduleProviders.end()) {
+    return nullptr;
+  }
+
+  return std::make_shared<TurboModuleImpl>(moduleName, callInvoker, it->second);
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/TurboModulesProvider.cpp:42]: >std::make_shared [](start = 9, length = 16)

Should we store the result to avoid creation for a second call?

File: vnext/Microsoft.ReactNative/TurboModulesProvider.h (hunk 4 above)
Comment (ZihanChen-MSFT) [reply vnext/Microsoft.ReactNative/TurboModulesProvider.h:20]: Is there any better choice? I need exactly a map here.

File: vnext/Microsoft.ReactNative/ReactPackageBuilder.h (hunk 2 above)
Comment (ZihanChen-MSFT) [reply vnext/Microsoft.ReactNative/ReactPackageBuilder.h:16]: This should be an internal class, since the idl file doesn't specify any constructor or implementing class.

File: vnext/Microsoft.ReactNative/TurboModulesProvider.cpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+std::vector<std::string> TurboModulesProvider::getEagerInitModuleNames() noexcept {
+  return {};
+}
+
+void TurboModulesProvider::AddModuleProvider(
```
Comment (dannyvv) [inline vnext/Microsoft.ReactNative/TurboModulesProvider.cpp:49]: Not sure what the android/ios semantics are here, but should we report an error if duplicate named modules are registered?

File: vnext/Microsoft.ReactNative/TurboModulesProvider.h (hunk 4 above)
Comment (asklar) [reply vnext/Microsoft.ReactNative/TurboModulesProvider.h:20]: @ZihanChen-MSFT a few options:
* `std::vector<std::pair<string, ReactModuleProvider>>`
* `f14map` (https://engineering.fb.com/developer-tools/f14/)
* `std::unordered_map` although if you expect to have less than 1000 entries, don't use a map

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp [hunk 7]
```diff
@@ -314,7 +314,7 @@ void ReactInstanceWin::Initialize() noexcept {
             auto instanceWrapper = facebook::react::CreateReactInstance(
                 std::string(), // bundleRootPath
                 std::move(cxxModules),
-                nullptr,
+                m_options.TurboModuleProvider,
```
Comment (acoates-ms) [inline vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp:317]: When verifying that we are able to run a TurboModule, you may find that they only work when not web-debugging.  I think we will probably need to add code to the instance to load the turbomodules as native modules when web debugging.

File: vnext/Microsoft.ReactNative/TurboModulesProvider.h (hunk 4 above)
Comment (ZihanChen-MSFT) [reply vnext/Microsoft.ReactNative/TurboModulesProvider.h:20]: I will use unordered_map for now, I don't quite believe we will have more than 1000 turbo modules running in an app.

File: vnext/Microsoft.ReactNative/TurboModulesProvider.cpp [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+                  method.Method(argReader, argWriter, nullptr, nullptr);
+                  return facebook::jsi::Value::undefined();
+                }
+                case MethodReturnType::Promise: {
+                  return facebook::react::createPromiseAsJSIValue(
+                      runtime, [=](facebook::jsi::Runtime &runtime, std::shared_ptr<facebook::react::Promise> promise) {
+                        method.Method(
+                            argReader,
+                            argWriter,
+                            [promise, &runtime](const IJSValueWriter &writer) {
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/TurboModulesProvider.cpp:102]: >&runtime [](start = 38, length = 8)

If we pass runtime like this, then the method must be synchronous. Otherwise, it will crash.

File: vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp (hunk 7 above)
Comment (ZihanChen-MSFT) [reply vnext/Microsoft.ReactNative/ReactHost/ReactInstanceWin.cpp]: Now Clipboard only registers to TurboModule when web debugger is disabled

File: vnext/Microsoft.ReactNative/Modules/ClipboardModule.cpp [hunk 9]
```diff
@@ -8,23 +8,37 @@
...
 
-/* static */ void Clipboard::setString(std::string content) noexcept {
-  winrt::Windows::ApplicationModel::DataTransfer::DataPackage data;
-  data.SetText(Microsoft::Common::Unicode::Utf8ToUtf16(content));
-  winrt::Windows::ApplicationModel::DataTransfer::Clipboard::SetContent(data);
+void Clipboard::getString(React::ReactPromise<React::JSValue> result) noexcept {
+  auto jsDispatcher = m_reactContext.JSDispatcher();
+  auto promise = result;
+  m_reactContext.UIDispatcher().Post(
+      [=]() -> winrt::fire_and_forget {
```
Comment (NickGerleman) [inline vnext/Microsoft.ReactNative/Modules/ClipboardModule.cpp:19]: Default capture by value is dangerous, and IIRC banned by CppCoreGuidelines which we try to follow. Consider explicitly adding to the capture group.

File: change/react-native-windows-2020-06-01-15-09-20-pull_request.json [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "none",
```
Comment (NickGerleman) [inline change/react-native-windows-2020-06-01-15-09-20-pull_request.json:2]: This should be "prerelease". We only want "none" in cases where the change doesn't affect the published package.

File: vnext/Microsoft.ReactNative/Modules/ClipboardModule.cpp [hunk 11]
```diff
@@ -6,25 +6,51 @@
...
+          break;
+        }
+        case AsyncStatus::Error: {
+          auto message = std::wstring(winrt::hresult_error(asyncOp.ErrorCode()).message());
+          jsDispatcher.Post([result, message] { result.Reject(message.c_str()); });
+          break;
+        }
+      }
+    });
+  } // namespace Microsoft::ReactNative
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/Modules/ClipboardModule.cpp:44]: >// namespace Microsoft::ReactNative [](start = 4, length = 35)

Please remove

File: vnext/Microsoft.ReactNative/Modules/ClipboardModule.cpp [hunk 12]
```diff
@@ -6,25 +6,51 @@
...
+  );
 }
 
-/* static */ void Clipboard::setString(std::string content) noexcept {
-  winrt::Windows::ApplicationModel::DataTransfer::DataPackage data;
-  data.SetText(Microsoft::Common::Unicode::Utf8ToUtf16(content));
-  winrt::Windows::ApplicationModel::DataTransfer::Clipboard::SetContent(data);
+void Clipboard::setString(std::string content) noexcept {
+  m_reactContext.UIDispatcher().Post([=] {
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/Modules/ClipboardModule.cpp:49]: >= [](start = 38, length = 1)

Please use explicit capture.

File: vnext/Microsoft.ReactNative/IReactPackageBuilder.idl [hunk 13]
```diff
@@ -15,4 +15,10 @@ namespace Microsoft.ReactNative {
     void AddModule(String moduleName, ReactModuleProvider moduleProvider);
     void AddViewManager(String viewManagerName, ReactViewManagerProvider viewManagerProvider);
   }
+
+  [webhosthidden]
+  interface IReactPackageBuilderExperimental
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/IReactPackageBuilder.idl:20]: >IReactPackageBuilderExperimental [](start = 12, length = 32)

We must not create new interfaces like this. Why the change is done comparing with the earlier version?
